src/Microsoft.ML.Core/Data/IEstimator.cs

```diff
@@ -0,0 +1,190 @@
+            public readonly VectorKind Kind;
+            public readonly DataKind ItemKind;
+            public readonly bool IsKey;
+            public readonly string[] MetadataKinds;
+
+            public Column(string name, VectorKind vecKind, DataKind itemKind, bool isKey, string[] metadataKinds)
+            {
+                Contracts.CheckNonEmpty(name, nameof(name));
+                Contracts.CheckValue(metadataKinds, nameof(metadataKinds));
+
+                Name = name;
+                Kind = vecKind;
+                ItemKind = itemKind;
+                IsKey = isKey;
+                MetadataKinds = metadataKinds;
+            }
+        }
+
+        public SchemaShape(Column[] columns)
+        {
+            Contracts.CheckValue(columns, nameof(columns));
+            Columns = columns;
+        }
+
+        /// <summary>
+        /// Create a schema shape out of the fully defined schema.
+        /// </summary>
+        public static SchemaShape Create(ISchema schema)
+        {
+            Contracts.CheckValue(schema, nameof(schema));
+            var cols = new List<Column>();
+
+            for (int iCol = 0; iCol < schema.ColumnCount; iCol++)
+            {
+                if (!schema.IsHidden(iCol))
+                {
+                    Column.VectorKind vecKind;
+                    var type = schema.GetColumnType(iCol);
+                    if (type.IsKnownSizeVector)
+                        vecKind = Column.VectorKind.Vector;
+                    else if (type.IsVector)
+                        vecKind = Column.VectorKind.VariableVector;
+                    else
+                        vecKind = Column.VectorKind.Scalar;
+
+                    var kind = type.ItemType.RawKind;
+                    var isKey = type.ItemType.IsKey;
+
+                    var metadataNames = schema.GetMetadataTypes(iCol)
+                        .Select(kvp => kvp.Key)
+                        .ToArray();
+                    cols.Add(new Column(schema.GetColumnName(iCol), vecKind, kind, isKey, metadataNames));
+                }
+            }
+            return new SchemaShape(cols.ToArray());
+        }
+
+        /// <summary>
+        /// Returns the column with a specified <paramref name="name"/>, and <c>null</c> if there is no such column.
+        /// </summary>
+        public Column FindColumn(string name)
+        {
+            Contracts.CheckValue(name, nameof(name));
+            return Columns.FirstOrDefault(x => x.Name == name);
+        }
+
+        // REVIEW: I think we should have an IsCompatible method to check if it's OK to use one schema shape
+        // as an input to another schema shape. I started writing, but realized that there's more than one way to check for
+        // the 'compatibility': as in, 'CAN be compatible' vs. 'WILL be compatible'.
+    }
+
+    /// <summary>
+    /// Exception class for schema validation errors.
+    /// </summary>
+    public class SchemaException : Exception
```

Comment:
>SchemaException [](start = 17, length = 15)

Since it's not part of contracts, who will ever throw it?

Reply:
We all shall, in time.

---
In reply to: [211687059](https://github.com/dotnet/machinelearning/pull/688#discussion_r211687059) [](ancestors = 211687059)